### PR TITLE
[Hub Generated] Review request for Microsoft.ContainerRegistry to add version preview/2019-12-01-preview

### PR DESCRIPTION
**Pull request by jikuma**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/jikuma/azure-rest-api-specs/dev-containerregistry-Microsoft.ContainerRegistry-2019-12-01-preview...master/).

### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/containerregistry/resource-manager/readme.md [hunk 1]
```diff
@@ -4,31 +4,40 @@
...
-
-
 ### Basic Information
+
 These are the global settings for the ContainerRegistry API.
 
 ``` yaml
 openapi-type: arm
-tag: package-2019-06-preview
+tag: package-preview-2019-12
```
Comment (raych1) [inline specification/containerregistry/resource-manager/readme.md]: How about follow the `yyyy-MM-preview` pattern for the tag?

Comment (jikuma) [reply specification/containerregistry/resource-manager/readme.md]: I do not remember changing this file, this file was changed by  "OpenApi Hub" on behalf of me.

**Comment by raych1** [comment]
@qiaozha , can you please take a look at the JS SDK failures?

**Comment by qiaozha** [comment]
@jikuma  This problem occurs because in your readme.md you've already set your default tag `package-preview-2019-12`. the input file for this tag is  `Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json`. Then in your readme.typescript.md file. you add 
`input-file:
  - Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json
  - Microsoft.ContainerRegistry/preview/2019-06-01-preview/containerregistry_build.json
  - Microsoft.ContainerRegistry/preview/2019-05-01-preview/containerregistry_scopemap.json`
However, both `Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json` and `Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json` have the same operation but the api-version parameter is different. the SDK generator don't know which one to choose. Any particular reason why you put this extra three files in readme.typescript.md?

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json [hunk 2]
```diff
@@ -0,0 +1,1291 @@
...
+            "name": "registryNameCheckRequest",
+            "in": "body",
+            "description": "The object containing information for the availability request.",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/RegistryNameCheckRequest"
+            }
+          }
+        ],
+        "responses": {
```
Comment (KrisBash) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: For all API responses in this spec, please add a "default" response to capture error cases. It should have a schema consistent with the [common error contract](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content).  Here is an example: https://github.com/Azure/azure-rest-api-specs/blob/0cf70260108cb40bc68d408674c187ed065f8c64/specification/batch/resource-manager/Microsoft.Batch/stable/2019-08-01/BatchManagement.json#L105

Even if this did not exist in prior versions, let's get it added going forward

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json [hunk 3]
```diff
@@ -0,0 +1,1291 @@
...
+          }
+        },
+        "x-ms-examples": {
+          "RegistryRegenerateCredential": {
+            "$ref": "./examples/RegistryRegenerateCredential.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerRegistry/registries/{registryName}/listUsages": {
```
Comment (KrisBash) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: Why is this "listUsages" and not /usages? It should follow the[ contract](https://armwiki.azurewebsites.net/api_contracts/UsagesAPIContract.html).

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json [hunk 4]
```diff
@@ -0,0 +1,1291 @@
+{
```
Comment (KrisBash) [inline specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: There are a number of nested resource types from the prior version not carried forward. Is that intentional? It's a suboptimal experience for the customer to produce a template to configure a resource if they have to use different API versions to configure different element/sub-resources.

**Review by KrisBash — CHANGES_REQUESTED** [review]
Please see comments

**Comment by jikuma** [comment]
> @jikuma This problem occurs because in your readme.md you've already set your default tag `package-preview-2019-12`. the input file for this tag is `Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json`. Then in your readme.typescript.md file. you add
> `input-file:
> 
> * Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json
> * Microsoft.ContainerRegistry/preview/2019-06-01-preview/containerregistry_build.json
> * Microsoft.ContainerRegistry/preview/2019-05-01-preview/containerregistry_scopemap.json`However, both`Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json`and`Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json` have the same operation but the api-version parameter is different. the SDK generator don't know which one to choose. Any particular reason why you put this extra three files in readme.typescript.md?

**Comment by jikuma** [comment]
@qiaozha  First, I did not make any change in readme files, these files were changed by OpenHub on behalf of me. 
Currently, there is a lot of operation shared between Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json and  Microsoft.ContainerRegistry/preview/2019-06-01-preview/containerregistry_build.json why is this not an issue now? Am I missing something?

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json (hunk 4 above)
Comment (jikuma) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: Added the deleted controllers back due to consistency.

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json (hunk 3 above)
Comment (jikuma) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: I do not want to make changes to the old controller as part of this PR. I will open WI to track these changes.

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json (hunk 2 above)
Comment (jikuma) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: The swagger file is auto-generated.  I do not think we should edit this file. I think we are missing some decorator on the controller for this. Can you point me to an example?

Secondly, I do not want to make changes to the old controller as part of this PR. I will open WI to track these changes.

File: specification/containerregistry/resource-manager/readme.md (hunk 1 above)
Comment (raych1) [reply specification/containerregistry/resource-manager/readme.md]: Thanks for the info. We will log a bug for `OpenApi Hub`.

**Comment by qiaozha** [comment]
> @qiaozha  First, I did not make any change in readme files, these files were changed by OpenHub on behalf of me.
> Currently, there is a lot of operation shared between Microsoft.ContainerRegistry/stable/2019-05-01/containerregistry.json and  Microsoft.ContainerRegistry/preview/2019-06-01-preview/containerregistry_build.json why is this not an issue now? Am I missing something?
> Also, how is all other languages like Java is able to generate SDK while it only fails for javascript?

First, I've checked this two files at tag 'package-2019-06-preview' of master which includes these two files. And I can't find any paths exsit in both files. maybe you can point it out for me?
Second, other languages sdks support mutlple-api while JavaScript SDK don't

File: specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json (hunk 3 above)
Comment (KrisBash) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: Can you help me understand why a new version (a preview too) is not a good opportunity to correct issues? I'm not sure what you mean by "old controller/api"

Comment (jikuma) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: changes the api and made it usages from listUsages

Comment (mnltejaswini) [reply specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json]: @KrisBash  I see that there is a contract to return the resource usages across Azure. But this API is not exactly the contract implementation as the payload is different and back then we didn't have this contract. Also, it loses its ability to call from ARM template if we remove **list** prefix. We need to evaluate our API usage and then introduce this change. Can we please take this change as a separate item and scope this version to customer managed key scenarios.

**Review by KrisBash — APPROVED** [review]
Discussed offline. Two issues must be fixed in a future update:
1) "default" response for all APIs
2) listUsages normalization to standard /usages contract

**Comment by raych1** [comment]
@ArcturusZhang , can you have a look at the SDK Go failures and suggest on the fix? Here's the error:
```
System.InvalidOperationException: codegen for preview swagger Microsoft.ContainerRegistry/preview/2019-12-01-preview/containerregistry.json must be under a preview subdirectory' 
```

File: specification/containerregistry/resource-manager/readme.go.md [hunk 5]
```diff
@@ -25,6 +26,15 @@ batch:
   - tag: package-2016-06-preview
 ```
 
+### Tag: package-2019-12-preview and go
+
+These settings apply only when `--tag=package-2019-12-preview --go` is specified on the command line.
+Please also specify `--go-sdk-folder=<path to the root directory of your azure-sdk-for-go clone>`.
+
+``` yaml $(tag) == 'package-2019-12-preview' && $(go)
+output-folder: $(go-sdk-folder)/services/$(namespace)/mgmt/2019-12-01-preview/$(namespace)
```
Comment (ArcturusZhang) [inline specification/containerregistry/resource-manager/readme.go.md]: Please change the output-folder to 
`$(go-sdk-folder)/services/preview/$(namespace)/mgmt/2019-12-01-preview/$(namespace)`

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]

**Comment by jikuma** [comment]
Now that all the checks are passing, can we merge this PR?

**Review by raych1 — APPROVED** [review]

**Review by ArcturusZhang — APPROVED** [review]

**Pull request merged**